### PR TITLE
Add Base implementation of website content resolving

### PR DESCRIPTION
**Pull request by Prokyonn**
builds up on the following merge requests:
* https://github.com/sulu/SuluContentBundle/pull/273
* https://github.com/sulu/SuluContentBundle/pull/272

### CODE REVIEW

File: Content/Application/ContentAggregator/ContentAggregator.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace Sulu\Bundle\ContentBundle\Content\Application\ContentAggregator;
+
+use Sulu\Bundle\ContentBundle\Content\Application\ContentMerger\ContentMergerInterface;
+use Sulu\Bundle\ContentBundle\Content\Domain\Exception\ContentNotFoundException;
+use Sulu\Bundle\ContentBundle\Content\Domain\Model\ContentRichEntityInterface;
+use Sulu\Bundle\ContentBundle\Content\Domain\Model\DimensionContentInterface;
+use Sulu\Bundle\ContentBundle\Content\Domain\Repository\DimensionContentRepositoryInterface;
+
+class ContentAggregator implements ContentAggregatorInterface
```
Comment (Prokyonn) [inline Content/Application/ContentAggregator/ContentAggregator.php]: This is the old `ContentResolver` I renamed it to `ContentAggregator` as it only "aggregates" or "combines" the data from multiple dimensions.

@alexander-schranz @chirimoya @wachterjohannes  WDYT about the naming?

Comment (alexander-schranz) [reply Content/Application/ContentAggregator/ContentAggregator.php]: I'm fine currently with that, but can we move it into an own merge request better for the git history.

Comment (Prokyonn) [reply Content/Application/ContentAggregator/ContentAggregator.php]: extracted to https://github.com/sulu/SuluContentBundle/pull/272

File: Content/Application/ContentObjects/ResolvableResource.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * (c) Sulu GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\ContentBundle\Content\Application\ContentObjects;
+
+class ResolvableResource
```
Comment (alexander-schranz) [inline Content/Application/ContentObjects/ResolvableResource.php]: At current state I would make this `@internal`

File: Content/Application/ContentObjects/ContentView.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+/*
+ * This file is part of Sulu.
+ *
+ * (c) Sulu GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\ContentBundle\Content\Application\ContentObjects;
```
Comment (alexander-schranz) [inline Content/Application/ContentObjects/ContentView.php]: eventually we could go with `Sulu\Bundle\ContentBundle\Content\Application\Content\Value;`

Avoiding `s` as it mostly singular.

File: Content/Application/ContentObjects/ResolvableResource.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+/*
+ * This file is part of Sulu.
+ *
+ * (c) Sulu GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\ContentBundle\Content\Application\ContentObjects;
```
Comment (alexander-schranz) [inline Content/Application/ContentObjects/ResolvableResource.php]: same as above

File: Content/Application/ContentResolver/ContentResolver.php [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+use Sulu\Bundle\ContentBundle\Content\Application\ContentObjects\ContentView;
+use Sulu\Bundle\ContentBundle\Content\Application\ContentObjects\ResolvableResource;
+use Sulu\Bundle\ContentBundle\Content\Application\ResourceLoader\ResourceLoaderProvider;
+use Sulu\Bundle\ContentBundle\Content\Application\TemplateResolver\TemplateResolverInterface;
+use Sulu\Bundle\ContentBundle\Content\Domain\Model\DimensionContentInterface;
+
+class ContentResolver implements ContentResolverInterface
+{
+    public function __construct(
+        private TemplateResolverInterface $templateResolver,
```
Comment (alexander-schranz) [inline Content/Application/ContentResolver/ContentResolver.php]: `iterable<ContentResolverInterface>`

File: Content/Application/ContentResolver/ContentResolver.php [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    {
+        $content = [];
+        $view = [];
+        $resolvableResources = [];
+
+        foreach ($contentViews as $name => $contentView) {
+            $result = $this->resolveContentView($contentView, $name);
+            $content = \array_merge($content, $result['content']);
+            $view = \array_merge($view, $result['view']);
+            $resolvableResources = \array_merge_recursive($resolvableResources, $result['resolvableResources']);
```
Comment (alexander-schranz) [inline Content/Application/ContentResolver/ContentResolver.php:86]: merge behave sometimes strange for nested arrays maybe we need to use `array_replace` / `array_recursive_replace`

File: Content/Application/PropertyResolver/MediaSelectionPropertyResolver.php [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+            ['ids' => $data['ids']],
+        );
+    }
+
+    public static function getType(): string
+    {
+        return 'media_selection';
+    }
+
+    public function getDefaultResourceLoader(): string
```
Comment (alexander-schranz) [inline Content/Application/PropertyResolver/MediaSelectionPropertyResolver.php]: Would I not create as a public method instead maybe a constant on the resolver itself.

File: Content/Application/PropertyResolver/MediaSelectionPropertyResolver.php [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+{
+    public function resolve(mixed $data, string $locale, array $params = []): ContentView
+    {
+        if (empty($data) || !\is_array($data) || !isset($data['ids'])) {
+            return ContentView::create([], ['ids' => []]);
+        }
+
+        return ContentView::createResolvables(
+            $data['ids'],
+            $params['resourceLoader'] ?? $this->getDefaultResourceLoader(),
```
Comment (alexander-schranz) [inline Content/Application/PropertyResolver/MediaSelectionPropertyResolver.php]:
```
$params['resourceLoader'] ?? MediaResourceLoader::NAME,
```

File: Content/Application/ResourceLoader/MediaResourceLoader.php [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+        $mappedResult = [];
+        foreach ($result as $media) {
+            $mappedResult[$media->getId()] = $media;
+        }
+
+        return $mappedResult;
+    }
+
+    public function getType(): string
```
Comment (alexander-schranz) [inline Content/Application/ResourceLoader/MediaResourceLoader.php]: Would be `static`

File: Content/Application/ResourceLoader/MediaResourceLoader.php [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        return 'media';
```
Comment (alexander-schranz) [inline Content/Application/ResourceLoader/MediaResourceLoader.php]: would provide a constant here to be used in the property resolver.

File: Content/Application/ResourceLoader/ResourceLoaderInterface.php [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+namespace Sulu\Bundle\ContentBundle\Content\Application\ResourceLoader;
+
+interface ResourceLoaderInterface
+{
+    /**
+     * @return array<int|string, object> index must be the ID of the object
+     */
+    public function load(array $ids, string $locale, array $params = []): array;
+
+    public function getType(): string;
```
Comment (alexander-schranz) [inline Content/Application/ResourceLoader/ResourceLoaderInterface.php]: static

File: Content/Application/TemplateResolver/TemplateResolver.php [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+use Sulu\Bundle\AdminBundle\Metadata\FormMetadata\StructureFormMetadataLoader;
+use Sulu\Bundle\AdminBundle\Metadata\FormMetadata\TypedFormMetadata;
+use Sulu\Bundle\ContentBundle\Content\Application\MetadataResolver\MetadataResolver;
+use Sulu\Bundle\ContentBundle\Content\Domain\Model\DimensionContentInterface;
+use Sulu\Bundle\ContentBundle\Content\Domain\Model\TemplateInterface;
+
+class TemplateResolver implements TemplateResolverInterface
+{
+    public function __construct(
+        private StructureFormMetadataLoader $structureFormMetadataLoader,
```
Comment (alexander-schranz) [inline Content/Application/TemplateResolver/TemplateResolver.php]: We should go over the FormMetadataProvider directly here or over the MetadataRegistry. Would avoid the structure service as it depends on the old factory which lives in the component namespace which should get be removed.

https://github.com/sulu/sulu/blob/1d04111a383c26e9c86c14b75721f1de11d668d5/src/Sulu/Bundle/AdminBundle/Metadata/FormMetadata/FormMetadataProvider.php#L49C41-L49C44

File: Content/Application/TemplateResolver/TemplateResolverInterface.php [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\ContentBundle\Content\Application\TemplateResolver;
+
+use Sulu\Bundle\ContentBundle\Content\Application\ContentObjects\ContentView;
+use Sulu\Bundle\ContentBundle\Content\Domain\Model\DimensionContentInterface;
+
+interface TemplateResolverInterface
```
Comment (alexander-schranz) [inline Content/Application/TemplateResolver/TemplateResolverInterface.php]: `ContentResolverInterface` would not create a seperate `TemplateResolverInterface`

File: Content/Application/TemplateResolver/TemplateResolver.php [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+/*
+ * This file is part of Sulu.
+ *
+ * (c) Sulu GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\ContentBundle\Content\Application\TemplateResolver;
```
Comment (alexander-schranz) [inline Content/Application/TemplateResolver/TemplateResolver.php]: move to `ContentResolver` namespace

File: Content/Application/ContentObjects/ContentView.php (hunk 3 above)
Comment (alexander-schranz) [reply Content/Application/ContentObjects/ContentView.php]: Maybe under `Sulu\Bundle\ContentBundle\Content\Application\ContentResolver\Value;`

File: Content/Application/MetadataResolver/MetadataResolver.php [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+     *
+     * @return ContentView[]
+     */
+    public function resolveItems(array $items, array $data, string $locale): array
+    {
+        $contentViews = [];
+        foreach ($items as $item) {
+            $name = $item->getName();
+            $type = $item->getType();
+            if ($item instanceof SectionMetadata || $item instanceof FormMetadata) {
```
Comment (alexander-schranz) [inline Content/Application/MetadataResolver/MetadataResolver.php]: check if there is not a method to get the flatted version of the metadata without the sections for resolving.

File: Content/Application/MetadataResolver/MetadataResolver.php [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+
+namespace Sulu\Bundle\ContentBundle\Content\Application\MetadataResolver;
+
+use Sulu\Bundle\AdminBundle\Metadata\FormMetadata\FormMetadata;
+use Sulu\Bundle\AdminBundle\Metadata\FormMetadata\ItemMetadata;
+use Sulu\Bundle\AdminBundle\Metadata\FormMetadata\SectionMetadata;
+use Sulu\Bundle\ContentBundle\Content\Application\ContentObjects\ContentView;
+use Sulu\Bundle\ContentBundle\Content\Application\PropertyResolver\PropertyResolverProvider;
+
+class MetadataResolver
```
Comment (alexander-schranz) [inline Content/Application/MetadataResolver/MetadataResolver.php:24]: if still required mark as `@internal`

File: Content/Application/PropertyResolver/MediaSelectionPropertyResolver.php [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\ContentBundle\Content\Application\PropertyResolver;
+
+use Sulu\Bundle\ContentBundle\Content\Application\ContentObjects\ContentView;
+
+class MediaSelectionPropertyResolver implements PropertyResolverInterface
```
Comment (alexander-schranz) [inline Content/Application/PropertyResolver/MediaSelectionPropertyResolver.php]: will be moved to media bundle.

Something like:

```php
if (\version_compare(\Composer\InstalledVersions::getVersion('sulu/sulu-content-bundle') ?? '0.0.0', '0.9', '>='
   && \version_compare(\Composer\InstalledVersions::getVersion('sulu/sulu-content-bundle') ?? '0.0.0', '0.10', '<'
) {
    $loader->load('content.xml'); // load property resolver service
}
```

File: Content/Application/ResourceLoader/MediaResourceLoader.php [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\ContentBundle\Content\Application\ResourceLoader;
+
+use Sulu\Bundle\MediaBundle\Media\Manager\MediaManagerInterface;
+
+class MediaResourceLoader implements ResourceLoaderInterface
```
Comment (alexander-schranz) [inline Content/Application/ResourceLoader/MediaResourceLoader.php]: same as for the resolver register only for content bundle 0.9

File: Content/UserInterface/Controller/Website/ContentController.php [hunk 19]
```diff
@@ -29,21 +30,25 @@
  */
 class ContentController extends AbstractController
 {
+    public function __construct(private ContentResolverInterface $contentResolver)
+    {
+    }
```
Comment (alexander-schranz) [inline Content/UserInterface/Controller/Website/ContentController.php]: Not sure if we should use here `getSubscribedServices`: 

https://github.com/sulu/sulu/blob/1d04111a383c26e9c86c14b75721f1de11d668d5/src/Sulu/Bundle/WebsiteBundle/Controller/WebsiteController.php#L182-L190

File: composer.json [hunk 20]
```diff
@@ -10,6 +10,12 @@
     ],
     "homepage": "https://github.com/sulu/SuluContentBundle",
     "license": "MIT",
+    "repositories": [
+        {
+            "type": "vcs",
+            "url": "https://github.com/Prokyonn/sulu.git"
+        }
+    ],
```
Comment (alexander-schranz) [inline composer.json:18]: think we need it the other way around?

Sulu need install this fork to test it 🤔

Or does this branch has something requirements / new classes we require from `sulu/sulu`?

File: Content/Application/ContentResolver/Value/ContentView.php [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+    /**
+     * @param mixed[] $view
+     */
+    public static function createResolvable(string|int $id, string $resourceLoaderKey, array $view): self
+    {
+        $resolvableResources = [
+            new ResolvableResource($id, $resourceLoaderKey),
+        ];
+
+        return new self($resolvableResources, $view);
```
Comment (alexander-schranz) [inline Content/Application/ContentResolver/Value/ContentView.php]:
```suggestion
        return new self(new ResolvableResource($id, $resourceLoaderKey), $view);
```

This returns an array but should not return an array as `content.media` with a single selection should directly be returned not as an array.

File: Content/Application/ResourceLoader/ResourceLoaderInterface.php [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace Sulu\Bundle\ContentBundle\Content\Application\ResourceLoader;
+
+interface ResourceLoaderInterface
+{
+    /**
+     * @param array<int|string> $ids
+     * @param mixed[] $params
+     *
+     * @return array<int|string, object> index must be the ID of the object
```
Comment (alexander-schranz) [inline Content/Application/ResourceLoader/ResourceLoaderInterface.php]:
```suggestion
     * @return array<int|string, mixed> index must be the ID of the object
```

The tag selection only returns the tag name and I think if somebody wants to return just a optimized array it should be possible so I would set this to mixed.

File: Content/Application/PropertyResolver/Resolver/BlockPropertyResolver.php [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+     * }[] $data
+     */
+    public function resolve(mixed $data, string $locale, array $params = []): ContentView
+    {
+        /** @var FieldMetadata $metadata */
+        $metadata = $params['metadata'];
+        $blockTypes = $metadata->getTypes();
+        $contentViews = [];
+        foreach ($data as $block) {
+            $type = $block['type'];
```
Comment (alexander-schranz) [inline Content/Application/PropertyResolver/Resolver/BlockPropertyResolver.php:62]: we should not forgot to implement block visitors into the content bundle
they are required for hide and scheduled blocks.

File: Content/Application/PropertyResolver/Resolver/BlockPropertyResolver.php [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+use Sulu\Bundle\ContentBundle\Content\Application\PropertyResolver\PropertyResolverInterface;
+
+class BlockPropertyResolver implements PropertyResolverInterface
+{
+    private MetadataResolver $metadataResolver;
+
+    /**
+     * Prevent circular dependency by injecting the MetadataResolver after instantiation.
+     */
+    public function setMetadataResolver(MetadataResolver $metadataResolver): void
```
Comment (alexander-schranz) [inline Content/Application/PropertyResolver/Resolver/BlockPropertyResolver.php:35]:
```suggestion
    /**
     * @internal 
     *
     * Prevent circular dependency by injecting the MetadataResolver after instantiation.
     */
    public function setMetadataResolver(MetadataResolver $metadataResolver): void
```

This method should not be part of the public API.

File: Content/Application/ContentResolver/Value/ContentView.php [hunk 25]
```diff
@@ -0,0 +1,90 @@
...
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\ContentBundle\Content\Application\ContentResolver\Value;
+
+/**
+ * @internal This class is intended for internal use only within the package/library. Modifying or depending on this class may result in unexpected behavior and is not supported.
+ */
+class ContentView
```
Comment (alexander-schranz) [inline Content/Application/ContentResolver/Value/ContentView.php:19]: As discussed we will require a `$resourceCallback` function as a parameter so the Link content type can after resolving set query and anchor ontop of it. This allows modification of the loaded resource by the propertyresolver later.

**Pull request merged**